@PrimaryColumn on Date field SQLite save issue

**Issue type:**

[ ] question
[x ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ x] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x ] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

I create an Entity with a primary column using Date as type, i can save using the save method but the next time the save method cant find the already saved entities with the same date because two problems:

* The search query of the save method uses ISO dates instead of the SQLite format` "YYYY-MM-DD"`, i resolve this creating a custom ValueTransformer and momentjs (but this must be a default behavior)

* The search query don't use the `date` function when searching for primary date column

The output of the save method in a second attempt looks like:

```bash
CONSOLE INFO file:///app/vendor.js:234819:21: query:  SELECT "UserPodcastHistory"."time" AS "UserPodcastHistory_time", "UserPodcastHistory"."userId" AS "UserPodcastHistory_userId" FROM "user_podcast_history" "UserPodcastHistory" WHERE (("UserPodcastHistory"."time" = ?) OR ("UserPodcastHistory"."time" = ?)) -- PARAMETERS: ["2018-10-09","2018-10-08"]
CONSOLE INFO file:///app/vendor.js:234819:21: query:  BEGIN TRANSACTION
CONSOLE INFO file:///app/vendor.js:234819:21: query:  INSERT INTO "user_podcast_history"("time", "userId") VALUES (?, NULL) -- PARAMETERS: ["2018-10-09 00:00:00.000"]
CONSOLE ERROR file:///app/vendor.js:234822:22: query failed:  INSERT INTO "user_podcast_history"("time", "userId") VALUES (?, NULL) -- PARAMETERS: ["2018-10-09 00:00:00.000"]
CONSOLE ERROR file:///app/vendor.js:234822:22: error:  SQLITE.ExecSQL Failed 0
CONSOLE INFO file:///app/vendor.js:234819:21: query:  ROLLBACK
CONSOLE ERROR file:///app/vendor.js:47035:24: ERROR QueryFailedError: SQLITE.ExecSQL Failed 0
```

--- a/src/driver/sqlite-abstract/AbstractSqliteDriver.ts
+++ b/src/driver/sqlite-abstract/AbstractSqliteDriver.ts
@@ -393,6 +393,11 @@ export abstract class AbstractSqliteDriver implements Driver {
 
             }
 
+            if (value instanceof Date) {
+                escapedParameters.push(DateUtils.mixedDateToUtcDatetimeString(value));
+                return this.createParameter(key, escapedParameters.length - 1);
+            }
+
             escapedParameters.push(value);
             return this.createParameter(key, escapedParameters.length - 1);
         }); // todo: make replace only in value statements, otherwise problems

--- a/src/driver/sqlite-abstract/AbstractSqliteDriver.ts
+++ b/src/driver/sqlite-abstract/AbstractSqliteDriver.ts
@@ -363,6 +363,10 @@ export abstract class AbstractSqliteDriver implements Driver {
                 return nativeParameters[key] === true ? 1 : 0;
             }
 
+            if (nativeParameters[key] instanceof Date) {
+                return DateUtils.mixedDateToUtcDatetimeString(nativeParameters[key]);
+            }
+
             return nativeParameters[key];
         });
 

--- a/test/github-issues/2286/entity/Example.ts
+++ b/test/github-issues/2286/entity/Example.ts
@@ -1,0 +1,12 @@
+import {Entity} from "../../../../src/decorator/entity/Entity";
+import {PrimaryColumn} from "../../../../src/decorator/columns/PrimaryColumn";
+import {Column} from "../../../../src/decorator/columns/Column";
+
+@Entity()
+export class Example {
+    @PrimaryColumn()
+    id: Date;
+
+    @Column()
+    text: string
+}

--- a/test/github-issues/2286/entity/Post.ts
+++ b/test/github-issues/2286/entity/Post.ts
@@ -1,0 +1,12 @@
+import {Entity} from "../../../../src/decorator/entity/Entity";
+import {PrimaryColumn} from "../../../../src/decorator/columns/PrimaryColumn";
+import {Column} from "../../../../src/decorator/columns/Column";
+
+@Entity()
+export class Post {
+    @PrimaryColumn("int")
+    id: number;
+
+    @Column()
+    dateTimeColumn: Date;
+} 

--- a/test/github-issues/2286/issue-2286.ts
+++ b/test/github-issues/2286/issue-2286.ts
@@ -3,20 +3,21 @@ import { createTestingConnections, closeTestingConnections, reloadTestingDatabas
 import { Connection } from "../../../src/connection/Connection";
 import { expect } from "chai";
 import { Post } from "./entity/Post";
+import { Example } from "./entity/Example";
 import { Between } from "../../../src";
 
 describe("github issues > #2286 find operators like MoreThan and LessThan doesn't work properly for date fields", () => {
 
     let connections: Connection[];
     before(async () => connections = await createTestingConnections({
-        entities: [__dirname + "/entity/*{.js,.ts}"],
+        entities: [ Post, Example ],
         schemaCreate: true,
         dropSchema: true,
         /* Test not eligible for better-sql where binding Dates is impossible */
         enabledDrivers: ["sqlite"]
     }));
     beforeEach(() => reloadTestingDatabases(connections));
-    
+
     after(() => closeTestingConnections(connections));
 
     it("should find a record by its datetime value with find options", () => Promise.all(connections.map(async connection => {
@@ -51,5 +52,29 @@ describe("github issues > #2286 find operators like MoreThan and LessThan doesn'
             .where("post.dateTimeColumn = :now", { now })
             .getOne();
         expect(postByDateEquals).to.not.be.undefined;
+    })));
+
+    it("should save, update, and load with a date PK", () => Promise.all(connections.map(async connection => {
+        const start = new Date("2000-01-01");
+        const middle = new Date("2000-06-30");
+        const end = new Date("2001-01-01");
+
+        await connection.manager.save(Example, { id: start, text: "start" });
+        await connection.manager.save(Example, { id: middle, text: "middle" });
+        await connection.manager.save(Example, { id: end, text: "end" });
+
+        const repo = connection.manager.getRepository(Example);
+
+        let example = await repo.findOneOrFail({ id: middle });
+
+        expect(example.text).to.be.equal("middle");
+
+        example.text = "in between";
+
+        await repo.save(example);
+
+        example = await repo.findOneOrFail({ id: middle });
+
+        expect(example.text).to.be.equal("in between");
     })));
 });

--- a/test/github-issues/2286/issue-2286.ts
+++ b/test/github-issues/2286/issue-2286.ts
@@ -1,0 +1,41 @@
+import "reflect-metadata";
+import { createTestingConnections, closeTestingConnections, reloadTestingDatabases } from "../../utils/test-utils";
+import { Connection } from "../../../src/connection/Connection";
+import { expect } from "chai";
+import { Post } from "./entity/Post";
+import { Between } from "../../../src";
+
+describe("github issues > #2286 find operators like MoreThan and LessThan doesn't work properly for date fields", () => {
+
+    let connections: Connection[];
+    before(async () => connections = await createTestingConnections({
+        entities: [__dirname + "/entity/*{.js,.ts}"],
+        schemaCreate: true,
+        dropSchema: true,
+        /* Test not eligible for better-sql where binding Dates is impossible */
+        enabledDrivers: ["sqlite"]
+    }));
+    beforeEach(() => reloadTestingDatabases(connections));
+    
+    after(() => closeTestingConnections(connections));
+
+    it("should find a record by its datetime value", () => Promise.all(connections.map(async connection => {
+        const start = new Date("2000-01-01");
+        const end = new Date("2001-01-01");
+        const middle = new Date("2000-06-30");
+        const post = new Post();
+        post.dateTimeColumn = middle;
+
+        await connection.manager.save(post);
+
+        const postByDateEquals = await connection.manager.findOne(Post, {
+            dateTimeColumn: middle
+        });
+        expect(postByDateEquals).to.not.be.undefined;
+
+        const postByDateBetween = await connection.manager.findOne(Post, {
+            dateTimeColumn: Between(start, end)
+        });
+        expect(postByDateBetween).to.not.be.undefined;
+    })));
+});

--- a/test/github-issues/2286/issue-2286.ts
+++ b/test/github-issues/2286/issue-2286.ts
@@ -19,7 +19,7 @@ describe("github issues > #2286 find operators like MoreThan and LessThan doesn'
     
     after(() => closeTestingConnections(connections));
 
-    it("should find a record by its datetime value", () => Promise.all(connections.map(async connection => {
+    it("should find a record by its datetime value with find options", () => Promise.all(connections.map(async connection => {
         const start = new Date("2000-01-01");
         const end = new Date("2001-01-01");
         const middle = new Date("2000-06-30");
@@ -37,5 +37,19 @@ describe("github issues > #2286 find operators like MoreThan and LessThan doesn'
             dateTimeColumn: Between(start, end)
         });
         expect(postByDateBetween).to.not.be.undefined;
+    })));
+
+    it("should find a record by its datetime value with query builder", () => Promise.all(connections.map(async connection => {
+        const now = new Date();
+        const post = new Post();
+        post.dateTimeColumn = now;
+
+        await connection.manager.save(post);
+
+        const postByDateEquals = await connection.manager.getRepository(Post)
+            .createQueryBuilder("post")
+            .where("post.dateTimeColumn = :now", { now })
+            .getOne();
+        expect(postByDateEquals).to.not.be.undefined;
     })));
 });

